### ISSUE TITLE
Nans appear in GPSolutionArray when a VectorVariable element is substituted for

### ISSUE THREAD

**Issue by pgkirsch** [convexopt/gpkit#269]
It seems that the solution isn't printing is because of the `nan's`. Notice that these `nan's` appear where the vector element value has been substituted, and it doesn't matter what the substituted value is.

So this issue has two parts:
- the vector variables have `nan's` where they should have the value of `eps`.
- the solution doesn't print when some of its values contain `nan's`

Code:
```
from gpkit.shortcuts import *
from numpy import ones

N = 6
L = 5.
dx = L/(N-1)

EI = Var("EI",10)

P = 100
p = Vec(N, "p", descr="Distributed load")
p = p.sub(p, P*ones(N))

V  = Vec(N, "V", descr="Internal shear")
M  = Vec(N, "M", descr="Internal moment")
th = Vec(N, "th", descr="Slope")
w  = Vec(N, "w", descr="Displacement")

eps = 1E-6 #something small

V[-1] = V[-1].sub(V[-1], eps)
M[-1] = M[-1].sub(M[-1], eps)
th[0] = th[0].sub(th[0], eps)
w[0]  = w[0].sub(w[0], eps)

objective = w[-1]

constraints = [EI*V.left[1:N]     >= EI*V[1:N]    + 0.5*dx*p.left[1:N]     + 0.5*dx*p[1:N],
               EI*M.left[1:N]     >= EI*M[1:N]    + 0.5*dx*V.left[1:N]     + 0.5*dx*V[1:N],
               EI*th.right[0:N-1] >= EI*th[0:N-1] + 0.5*dx*M.right[0:N-1]  + 0.5*dx*M[0:N-1],
               EI*w.right[0:N-1]  >= EI*w[0:N-1]  + 0.5*dx*th.right[0:N-1] + 0.5*dx*th[0:N-1]]

gp = GP(objective, constraints)

sol = gp.solve()

print sol.table()
```

Output:
```
In [17]: %run beam.py
Using solver 'mosek'
Solving for 20 variables.
Solving took 0.0283 seconds.

Cost
 0.781   

Variables
---------
EI : 10      

Constant sensitivities
----------------------
EI : -4  
```

If I inspect `sol`, the "second" solution seems to be:
```
'variables': {EI: array(10),
  w: array([        nan,  0.05125112,  0.18500127,  0.36625144,  0.57000162,
          0.78125184]),
  M: array([ 12.5000031 ,   8.00000206,   4.50000166,   2.00000138,
           0.50000117,          nan]),
  V: array([ 50.00000892,  40.00000314,  30.00000193,  20.00000143,
          10.00000115,          nan]),
  th: array([        nan,  1.02500128,  1.65000149,  1.97500168,  2.10000187,
          2.12500215])}
```

**Comment by whoburg** [comment]
Smaller minimum working example:

```python
from gpkit import Variable, VectorVariable, GP
x = VectorVariable(5, 'x')
x[0] = x[0].sub(x[0], 1e-3)
prob = GP(sum(x), [x[1:] >= 1])
sol = prob.solve()
print sol.table()
sol["variables"]
```

Output:
```python

Using solver 'cvxopt'
Solving for 4 variables.
Solving took 0.00616 seconds.

Cost
----
 4.001

Free variables
--------------

Swept variables
---------------

Constants
---------

Constant and swept variable sensitivities
-----------------------------------------
```

**Comment by pgkirsch** [comment]
Will work on this today

**Comment by whoburg** [comment]
Not that you need advice, but I'd probably start by inspecting `x[0]` after the sub, as well as `sum(x)`.

**Comment by pgkirsch** [comment]
The problem arises due to `_parse_result` and I am working on a fix. But I have a question about the end goal here.

Consider the following GP. It is equivalent to the MWE:
```
from gpkit import Variable, GP 
 
x0 = Variable('x0') 
x1 = Variable('x1') 
x2 = Variable('x2') 
x3 = Variable('x3') 
x4 = Variable('x4') 
 
x0 = x0.sub(1e-3) 
 
gp = GP(x0 +  x1 + x2 + x3 + x4, [x1 >= 1., x2 >= 1., x3 >= 1., x4 >= 1.]) 
sol = gp.solve() 
print sol.table() 
```

When you solve this, it will print a solution that looks like:

```
Free variables
--------------
x1 : 1 
x2 : 1 
x3 : 1 
x4 : 1 
```

My design question is whether the VectorVariable solution to the MWE should print as follows:

```
Free variables
--------------
x : [ 0.001     1         1         1        ... ] 
```

This is ever so slightly different and inaccurate because the first element is not free, but also makes the most sense from the "Vector Variable" perspective.

**Comment by bqpd** [comment]
What if...there was a plain "mixed variables" category for such edge cases?
Ugh, that also seems gross.

**Comment by pgkirsch** [comment]
Yeah I agree that's kind of nasty. What about an asterisk to denote that a particular element is fixed?

**Comment by bqpd** [comment]
So, just display nans as asteriks?

**Comment by pgkirsch** [comment]
Err.. I guess that could work? But I meant something like:

```
Free variables
--------------
x : [ 0.001*     1         1         1        ... ] 
```

*values denoted with an asterisk are fixed

Or something.

**Comment by bqpd** [comment]
But how do we tell between a partly-substituted free variable and a
partly-free constant? :p

**Comment by pgkirsch** [comment]
Valid.

By the way, it's unclear to me: does `_parse_result()` see the substituted element at all? If so, I can't seem to find where.

**Comment by whoburg** [comment]
The design issue is a fair point, and I don't have strong feelings about the right solution. A simple solution, which I'm fine with, is printing any `VectorVariable` as a free variable if it has any free variables inside it. But a more complicated scheme like the asterisks is also fine with me.

I think you guys are a step ahead of me -- is it clear what's introducing the `nan`s?

**Comment by pgkirsch** [comment]
Yes, sorry, it's the following excerpt of code within `_parse_result()`:

```
try:
    variables[veckey][idx] = variables[var]
    sensitivities["variables"][veckey][var.descr["idx"]] = \
                        sensitivities["variables"][var]
except KeyError:
    variables[veckey] = np.empty(var.descr["shape"]) + np.nan
    sensitivities["variables"][veckey] = (
                        np.empty(var.descr["shape"]) + np.nan)
```

This code is in a loop that essentially takes
```
variables = {x1: 1.0, x2: 1.0,... x4: 1.0}
```
and transforms it into
```
variables = {x = [1.0, 1.0, 1.0, 1.0]}
```
That particular segment of code is creating the nan-initialized list for `x`, if it doesn't exist already. 

The problem is that it knows that `x` should have 5 elements, but only knows what 4 of them are (see the original `variables`, so what it is actually producing, as we have seen is 
```
variables = {x = [nan, 1.0, 1.0, 1.0, 1.0]}
```

**Comment by pgkirsch** [comment]
Please can someone explain the following segment of code to me:
```
         # var does not exist yet 
         for var, S in sensitivities["variables"].items():
             if var not in self.substitutions and abs(S) > 1e-4:
                 raise RuntimeWarning("free variable too sensitive:"
                                     " S_{%s} = %0.2e" % (var, S))
```

If I add a couple print statements afterwards, I get the following output:
```
         print sensitivities["variables"].items()
         # [(x_(4,), 0.0), (x_(1,), 0.0), (x_(2,), 0.0), (x_(3,), 0.0)]
         print var.descr["shape"]
         # (5,)
```

How is `var.descr` being defined?

**Comment by pgkirsch** [comment]
Is an acceptable solution just to have the following output:

```
Free variables
--------------
x_(1,) : 1 
x_(2,) : 1 
x_(3,) : 1 
x_(4,) : 1 
```

(ignores substituted element entirely)

**Comment by pgkirsch** [comment]
I have identified why the solution doesn't print when it contains `NaN`s.

This is due to `minval` in `small_scripts.results_table()`. Specifically, the following line causes the issue:

Removing the `if` statement resolves this issue. However, `minval` clearly serves a purpose, so I propose adding an `or` statement that allows `nan`s to pass through. This would fix half of this issue. How would you feel about this @whoburg, @bqpd ?

**Comment by pgkirsch** [comment]
I have identified why the solution doesn't print when it contains `NaN`s.

This is due to `minval` in `small_scripts.results_table()`. Specifically, the following line causes the issue:

```
decorated = [(bool(v.shape) if isinstance(v, Iterable) else False,
                 (varfmt % k),
                 i, k, v) for i, (k, v) in enumerate(data.items())
                 if np.max(abs(v)) >= minval]
```

Removing the `if` statement resolves this issue. 

```
decorated = [(bool(v.shape) if isinstance(v, Iterable) else False,
                 (varfmt % k),
                 i, k, v) for i, (k, v) in enumerate(data.items())]
```

However, `minval` clearly serves a purpose, so I propose modifying the conditional so that it specifically allows `nan`s to pass through. This would fix half of this issue. How would you feel about this @whoburg, @bqpd ?

**Comment by pgkirsch** [comment]
Also, I know it's not my place to criticize someone else's code, but it physically hurts me to read that line.
It's like everything bad about python in 4 nasty, ultra-compressed lines.

**Comment by bqpd** [comment]
Hahaha. I can laugh because that's one of @whoburg's fixes to the terrible mess that came before.

**Comment by bqpd** [comment]
You could do this with `np.max(abs(v)) >= minval or isnan(np.max(abs(v)))`, I think...

**Comment by pgkirsch** [comment]
yup done, I'm just creating a pull request now

**Comment by pgkirsch** [comment]
Interesting discover about the MWE.

Code 1 (unmodified MWE):
```
from gpkit import VectorVariable, GP
x = VectorVariable(5, 'x')
x[0] = x[0].sub(x[0], 1e-3)
prob = GP(sum(x), [x[1:] >= 1])
sol = prob.solve()
print sol.table()
```

Output 1:
```
Cost
----
 4.001

Free variables
--------------
x : [ nan       1         1         1        ... ] 

Swept variables
---------------

Constants
---------

Constant and swept variable sensitivities
-----------------------------------------
x : [ nan       0         0         0        ... ] 
```

Code 2:
```
from gpkit import VectorVariable, GP
x = VectorVariable(5, 'x')
prob = GP(sum(x), [x[1:] >= 1], {x[0]: 1e-3})
sol = prob.solve()
print sol.table()
```

Output 2:
```
Cost
----
 4.001

Free variables
--------------
x : [ 0.001     1         1         1        ... ] 

Swept variables
---------------

Constants
---------

Constant and swept variable sensitivities
-----------------------------------------
```

My conclusion is therefore that the problem isn't in `geometric_program` as I thought all this time, but rather in the `Variable.sub()` method.

**Comment by pgkirsch** [comment]
This issue is also true for a non-vector-variable case:

```
x0 = Variable('x0')
x1 = Variable('x1')
x2 = Variable('x2')
x3 = Variable('x3')
x4 = Variable('x4')

x0 = x0.sub(1e-3)

gp = GP(x0 +  x1 + x2 + x3 + x4, [x1 >= 1., x2 >= 1., x3 >= 1., x4 >= 1.])#, {x0: 1e-3})
sol = gp.solve()
print sol.table()
```

```
Cost
----
 4.001

Free variables
--------------
x1 : 1 
x2 : 1 
x3 : 1 
x4 : 1 

Swept variables
---------------

Constants
---------

Constant and swept variable sensitivities
-----------------------------------------
```


vs.

```
x0 = Variable('x0')
x1 = Variable('x1')
x2 = Variable('x2')
x3 = Variable('x3')
x4 = Variable('x4')

#x0 = x0.sub(1e-3)

gp = GP(x0 +  x1 + x2 + x3 + x4, [x1 >= 1., x2 >= 1., x3 >= 1., x4 >= 1.], {x0: 1e-3})
sol = gp.solve()
print sol.table()
```

```
Cost
----
 4.001

Free variables
--------------
x1 : 1 
x2 : 1 
x3 : 1 
x4 : 1 

Swept variables
---------------

Constants
---------
x0 : 0.001 

Constant and swept variable sensitivities
-----------------------------------------
```

**Comment by pgkirsch** [comment]
In fact the following code is also different:

```
x0 = Variable('x0', value=1e-3)
x1 = Variable('x1')
x2 = Variable('x2')
x3 = Variable('x3')
x4 = Variable('x4')

gp = GP(x0 +  x1 + x2 + x3 + x4, [x1 >= 1., x2 >= 1., x3 >= 1., x4 >= 1.])
sol = gp.solve()
print sol.table()
```

```
Cost
----
 4.001

Free variables
--------------
x1 : 1 
x2 : 1 
x3 : 1 
x4 : 1 

Swept variables
---------------

Constants
---------
x0 : 0.001 

Constant and swept variable sensitivities
-----------------------------------------
```


This seems to contradict the doctstring in `Variable.sub()` that claims
```
assert x.sub(3) == Variable('x', value=3)
```

**Comment by bqpd** [comment]
That docstring is wrong, sorry. I think it may have come from an earlier design discussion @whoburg and I had.

**Comment by pgkirsch** [comment]
I think I'm close to fixing the remainder of this issue.

Essentially, because the `Variable.sub()` method is returning a `Monomial` instead of a `Variable`, the objective function is then passed to `GP()` as `0.001 + x1 + ...` instead of `x0 + x1 + ...`, which eventually leads to the `parse_results()` missing `x0` entirely.

Unfortunately, I don't think I know of an elegant way of fixing `Variable.sub`. I want to do something really simple like `self.desc["value"] = arg`, but this breaks down for vector variables.

Any and all suggestions are appreciated! @whoburg @bqpd

Design question: do we even need 3 separate ways of performing substitutions?

**Comment by whoburg** [comment]
This is a great example of a really excellent discussion / workflow! Thanks @pgkirsch for helping to nail down the issue(s).

In my opinion the important issue is not how many ways of performing substitution we have, but rather that they all have the same behavior/syntax/api. This seems like an example that breaks that, since passing a substitution dict into the GP worked, but substituting the variable beforehand didn't.

This is definitely related to another ticket where @bqpd and I were discussing what `x.sub(3)` should do -- need to track that down.

**Comment by whoburg** [comment]
The related issue is #267.

@bqpd, any thoughts? Is `sub` perhaps not what users should be using to say that a `Variable` (and especially an element of a `VectorVariable`) has a value? Or is this a bug / design flaw?

**Comment by bqpd** [comment]
I think Variable.sub should fall back on Signomial.sub as part of the Signomial inheritance contract. If something's a Signomial, subbing into it replaces varkeys instead of setting their values.

**Comment by pgkirsch** [comment]
I'm confused because isn't that already what happens?

**Comment by bqpd** [comment]
Yeah, sorry for the confusion. I was answering @whoburg's "Is sub perhaps not what users should be using to say that a Variable (and especially an element of a VectorVariable) has a value?" with the main reason Yes.

**Comment by whoburg** [comment]
Ok, I'm fine with `Variable.sub` falling back on `Signomial.sub`. And it seems indeed that `VectorVariable` is handling `sub` in a reasonable way:

```
In [1]: from gpkit import VectorVariable

In [2]: x = VectorVariable(3, 'x')

In [3]: x
Out[3]: gpkit.PosyArray([x_(0,), x_(1,), x_(2,)])

In [4]: x[0] = x[0].sub(4)

In [5]: x
Out[5]: gpkit.PosyArray([4, x_(1,), x_(2,)])

In [6]: x.sum()
Out[6]: gpkit.Posynomial(4 + x_(1,) + x_(2,))
```

It seems to me the place things start to go wrong is the `nan`s in the `GPSolutionArray`. From a user's perspective, it seems the natural design would be for substituted constants in a `VectorVariable` to also appear in the solution array. Does that break or design philosophy or inheritance or substitution in some fundamental way?

A weird edge case is a fully-substituted `VectorVariable`, which of course no longer contains `VarKeys`...

**Comment by whoburg** [comment]
Ok, thanks to the fixes so far (1238df57a7059f390a9c332bb1e10eb75d24f4a1), the solution to the original problem now prints, with nans.

I'm changing the title and moving this to 0.3.0. Next steps are

 - [ ] Decide whether a `GPSolutionArray` should ever contain nans.
 - [ ] Recommend best practices for setting elements of a `VectorVariable` to a constant

**Comment by bqpd** [comment]
I think that a GPSolutionArray should contain nans to indicate missing values: if you try to e.g. do vector multiplication with an incomplete vector, the nans will propagate through nicely.

Best practices is definitely a substitution dictionary, as below:

```python
from gpkit.shortcuts import *
from numpy import ones

N = 6
L = 5.
dx = L/(N-1)

EI = Var("EI",10)

P = 100
p = Vec(N, "p", descr="Distributed load")
p = p.sub(p, P*ones(N))

V  = Vec(N, "V", label="Internal shear")
M  = Vec(N, "M", label="Internal moment")
th = Vec(N, "th", label="Slope")
w  = Vec(N, "w", label="Displacement")

eps = 1E-6 #something small

substitutions = {var: eps for var in [V[-1], M[-1], th[0], w[0]]}

objective = w[-1]

constraints = [EI*V.left[1:N]     >= EI*V[1:N]    + 0.5*dx*p.left[1:N]     + 0.5*dx*p[1:N],
               EI*M.left[1:N]     >= EI*M[1:N]    + 0.5*dx*V.left[1:N]     + 0.5*dx*V[1:N],
               EI*th.right[0:N-1] >= EI*th[0:N-1] + 0.5*dx*M.right[0:N-1]  + 0.5*dx*M[0:N-1],
               EI*w.right[0:N-1]  >= EI*w[0:N-1]  + 0.5*dx*th.right[0:N-1] + 0.5*dx*th[0:N-1]]

m = Model(objective, constraints, substitutions)

sol = m.solve()

print sol.table()
```
